From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Andr=C3=A9=20Zwing?= <nerv@dawncrow.de>
Date: Tue, 21 Feb 2023 21:39:02 +0100
Subject: server: Allow ARM on x86-64

---
 server/registry.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/server/registry.c b/server/registry.c
index cc9a33fff1d..97288c8d1f2 100644
--- a/server/registry.c
+++ b/server/registry.c
@@ -1850,7 +1850,11 @@ static void init_supported_machines(void)
 #ifdef __i386__
     if (prefix_type == PREFIX_32BIT) supported_machines[count++] = IMAGE_FILE_MACHINE_I386;
 #elif defined(__x86_64__)
-    if (prefix_type == PREFIX_64BIT) supported_machines[count++] = IMAGE_FILE_MACHINE_AMD64;
+    if (prefix_type == PREFIX_64BIT)
+    {
+        supported_machines[count++] = IMAGE_FILE_MACHINE_AMD64;
+        supported_machines[count++] = IMAGE_FILE_MACHINE_ARMNT;
+    }
     supported_machines[count++] = IMAGE_FILE_MACHINE_I386;
 #elif defined(__arm__)
     if (prefix_type == PREFIX_32BIT) supported_machines[count++] = IMAGE_FILE_MACHINE_ARMNT;
