From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Andr=C3=A9=20Zwing?= <nerv@dawncrow.de>
Date: Mon, 4 Sep 2023 19:53:14 +0200
Subject: ntdll: Improve the locale hack by applying a hack

---
 dlls/ntdll/loader.c | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index 60de59543c8..c00ea797eac 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -4396,6 +4396,11 @@ void loader_init( CONTEXT *context, void **entry )
         ANSI_STRING ctrl_routine = RTL_CONSTANT_STRING( "CtrlRoutine" );
         WINE_MODREF *kernel32;
         PEB *peb = NtCurrentTeb()->Peb;
+        EVENT_BASIC_INFORMATION ebi;
+        OBJECT_ATTRIBUTES attr;
+        UNICODE_STRING nameW;
+        BOOL hack = FALSE;
+        HANDLE event;
 
         peb->LdrData            = &ldr;
         peb->FastPebLock        = &peb_lock;
@@ -4422,6 +4427,21 @@ void loader_init( CONTEXT *context, void **entry )
         if (!default_load_path)
             get_dll_load_path( peb->ProcessParameters->ImagePathName.Buffer, NULL, dll_safe_mode, &default_load_path );
 
+        RtlInitUnicodeString( &nameW, L"\\KernelObjects\\__wineboot_event" );
+        InitializeObjectAttributes( &attr, &nameW, 0, 0, NULL );
+        status = NtOpenEvent( &event, EVENT_ALL_ACCESS, &attr );
+        if (NT_SUCCESS(status))
+            status = NtQueryEvent( event, EventBasicInformation, &ebi, sizeof(ebi), NULL );
+ 
+        if (NT_SUCCESS(status) && !ebi.EventState)
+        {
+            if (NtCurrentTeb()->WowTebOffset) WARN("Activated a hack for the hack\n");
+            hack = TRUE;
+        }
+        else if (NtCurrentTeb()->WowTebOffset) TRACE("Disabled a hack for the hack\n");
+ 
+        if (hack && NtCurrentTeb()->WowTebOffset) init_wow64( context );
+
         wm = build_main_module();
         build_ntdll_module();
 #ifdef __arm64ec__
@@ -4444,7 +4464,7 @@ void loader_init( CONTEXT *context, void **entry )
         if (needs_elevation())
             elevate_token();
 
-        if (NtCurrentTeb()->WowTebOffset) init_wow64( context );
+        if (!hack && NtCurrentTeb()->WowTebOffset) init_wow64( context );
 
         get_env_var( L"WINESYSTEMDLLPATH", 0, &system_dll_path );
         if (wm->ldr.Flags & LDR_COR_ILONLY)
