From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Andr=C3=A9=20Zwing?= <nerv@dawncrow.de>
Date: Fri, 1 Mar 2024 20:53:21 +0100
Subject: ntdll: Enable the Hack for some internal applications

---
 dlls/ntdll/loader.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index c00ea797eac..b8d4bc3b135 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -4438,6 +4438,15 @@ void loader_init( CONTEXT *context, void **entry )
             if (NtCurrentTeb()->WowTebOffset) WARN("Activated a hack for the hack\n");
             hack = TRUE;
         }
+        else if (!wcsicmp(NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer+1, L":\\windows\\syswow64\\winedbg.exe") ||
+                 !wcsicmp(NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer+1, L":\\windows\\syswow64\\cmd.exe") ||
+                 !wcsicmp(NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer+1, L":\\windows\\syswow64\\notepad.exe") ||
+                 !wcsicmp(NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer+1, L":\\windows\\syswow64\\regsvr32.exe") ||
+                 !wcsicmp(NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer+1, L":\\windows\\syswow64\\regedit.exe"))
+        {
+            if (NtCurrentTeb()->WowTebOffset) WARN("Activated a hack for the hack\n");
+            hack = TRUE;
+        }
         else if (NtCurrentTeb()->WowTebOffset) TRACE("Disabled a hack for the hack\n");
  
         if (hack && NtCurrentTeb()->WowTebOffset) init_wow64( context );
