From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Andr=C3=A9=20Zwing?= <nerv@dawncrow.de>
Date: Sun, 13 Oct 2024 11:57:12 +0200
Subject: ntdll: Refuse to run on page sizes other than 4k

---
 dlls/ntdll/unix/server.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/ntdll/unix/server.c b/dlls/ntdll/unix/server.c
index f03d008e2ae..ef7aade9c13 100644
--- a/dlls/ntdll/unix/server.c
+++ b/dlls/ntdll/unix/server.c
@@ -1257,6 +1257,11 @@ static int setup_config_dir(void)
     struct stat st;
     int fd_cwd = open( ".", O_RDONLY );
 
+#ifdef _SC_PAGESIZE
+    if (sysconf( _SC_PAGESIZE ) != 4096)
+        fatal_error("You are running with the wrong page size: %ld!\nPlease consult https://github.com/AndreRH/hangover/wiki/Page-Size\n", sysconf( _SC_PAGESIZE ));
+#endif
+
     if (chdir( config_dir ) == -1)
     {
         if (errno != ENOENT) fatal_perror( "cannot use directory %s", config_dir );
