From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Billy Laws <blaws05@gmail.com>
Date: Fri, 27 Sep 2024 11:29:21 +0000
Subject: HACK: define some extra arm64ec symbols to workaround llvm bugs

---
 aclocal.m4 |   6 ++
 configure  | 210 +++++++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 216 insertions(+)

diff --git a/aclocal.m4 b/aclocal.m4
index b319e283209..4eae9d22999 100644
--- a/aclocal.m4
+++ b/aclocal.m4
@@ -198,6 +198,12 @@ ac_wine_try_cflags_saved_exeext=$ac_exeext
 CFLAGS="$CFLAGS -nostdlib -nodefaultlibs $1"
 ac_exeext=".exe"
 AC_LINK_IFELSE([AC_LANG_SOURCE([[void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }]])],
                [AS_VAR_SET(ac_var,yes)], [AS_VAR_SET(ac_var,no)])
 CFLAGS=$ac_wine_try_cflags_saved
diff --git a/configure b/configure
index 64e0df59280..70b467cb734 100755
--- a/configure
+++ b/configure
@@ -10258,6 +10258,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10293,6 +10299,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10495,6 +10507,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10529,6 +10547,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10563,6 +10587,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10597,6 +10627,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10631,6 +10667,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10665,6 +10707,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10699,6 +10747,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10733,6 +10787,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10767,6 +10827,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10801,6 +10867,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10835,6 +10907,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10869,6 +10947,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10903,6 +10987,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10937,6 +11027,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -10971,6 +11067,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11005,6 +11107,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11039,6 +11147,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11073,6 +11187,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11107,6 +11227,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11141,6 +11267,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11175,6 +11307,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11211,6 +11349,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11247,6 +11391,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11281,6 +11431,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11315,6 +11471,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11349,6 +11511,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11383,6 +11551,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11417,6 +11591,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11451,6 +11631,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11508,6 +11694,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11542,6 +11734,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11580,6 +11778,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
@@ -11614,6 +11818,12 @@ ac_exeext=".exe"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 void *__os_arm64x_dispatch_ret = 0;
+void *__os_arm64x_dispatch_call_no_redirect = 0;
+void *memcpy() {};
+void *__os_arm64x_check_icall =0;
+void *memset() {};
+void *memmove() {};
+
 int __cdecl mainCRTStartup(void) { return 0; }
 _ACEOF
 if ac_fn_c_try_link "$LINENO"
