From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Andr=C3=A9=20Zwing?= <nerv@dawncrow.de>
Date: Sat, 19 Oct 2024 19:03:48 +0200
Subject: HACK: Don't build box64 library for ARM64EC

---
 configure    | 1 +
 configure.ac | 1 +
 2 files changed, 2 insertions(+)

diff --git a/configure b/configure
index 70b467cb734..c0187c3abda 100755
--- a/configure
+++ b/configure
@@ -20967,6 +20967,7 @@ enable_vcruntime140_1=${enable_vcruntime140_1:-x86_64,arm64ec}
 if test -n "$PE_ARCHS"
 then
     enable_box64cpu=${enable_box64cpu:-aarch64}
+    as_fn_append arm64ec_DISABLED_SUBDIRS " libs/box64"
     enable_wow64=${enable_wow64:-aarch64,x86_64}
     enable_wow64win=${enable_wow64win:-aarch64,x86_64}
     enable_wow64cpu=${enable_wow64cpu:-x86_64}
diff --git a/configure.ac b/configure.ac
index 531924844df..5a8bda9a9a8 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2338,6 +2338,7 @@ enable_vcruntime140_1=${enable_vcruntime140_1:-x86_64,arm64ec}
 if test -n "$PE_ARCHS"
 then
     enable_box64cpu=${enable_box64cpu:-aarch64}
+    AS_VAR_APPEND([arm64ec_DISABLED_SUBDIRS],[" libs/box64"])
     enable_wow64=${enable_wow64:-aarch64,x86_64}
     enable_wow64win=${enable_wow64win:-aarch64,x86_64}
     enable_wow64cpu=${enable_wow64cpu:-x86_64}
