diff --git a/dlls/winepulse.drv/pulse.c b/dlls/winepulse.drv/pulse.c
index 16e24cf..25a31a3 100644
--- a/dlls/winepulse.drv/pulse.c
+++ b/dlls/winepulse.drv/pulse.c
@@ -245,8 +245,9 @@ static NTSTATUS pulse_process_attach(void *args)
     pthread_mutexattr_t attr;
 
     pthread_mutexattr_init(&attr);
+#ifndef __ANDROID__    
     pthread_mutexattr_setprotocol(&attr, PTHREAD_PRIO_INHERIT);
-
+#endif
     if (pthread_mutex_init(&pulse_mutex, &attr) != 0)
         pthread_mutex_init(&pulse_mutex, NULL);
 
