From d8d66fbde9b3baa97313d1de3d40a2a7f2bd0a53 Mon Sep 17 00:00:00 2001
From: pipetto-crypto <pipetto-crypto@gmail.com>
Date: Mon, 2 Jun 2025 18:49:26 +0200
Subject: [PATCH] winex11: force GLX extension through env var

---
 dlls/winex11.drv/opengl.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/winex11.drv/opengl.c b/dlls/winex11.drv/opengl.c
index 8ab3e3474d3..f8d444cfa9b 100644
--- a/dlls/winex11.drv/opengl.c
+++ b/dlls/winex11.drv/opengl.c
@@ -542,6 +542,7 @@ static void init_opengl(void)
 {
     int error_base, event_base;
     unsigned int i;
+    static int wine_x11forceglx = -1;
 
     /* No need to load any other libraries as according to the ABI, libGL should be self-sufficient
        and include all dependencies */
@@ -630,7 +631,10 @@ static void init_opengl(void)
 
     if(!X11DRV_WineGL_InitOpenglInfo()) goto failed;
 
-    if (XQueryExtension( gdi_display, "GLX", &glx_opcode, &event_base, &error_base ))
+    if (wine_x11forceglx == -1)
+    	wine_x11forceglx = getenv("WINE_X11FORCEGLX") && atoi(getenv("WINE_X11FORCEGLX"));
+
+    if (XQueryExtension( gdi_display, "GLX", &glx_opcode, &event_base, &error_base ) || wine_x11forceglx)
     {
         TRACE("GLX is up and running error_base = %d\n", error_base);
     } else {
-- 
2.53.0.windows.2

